plugins/scenario/utils/TokenSourcer.ts

```diff
@@ -15,6 +14,40 @@ interface SourceTokenParameters {
   blacklist: string[] | undefined;
 }
 
+export async function fetchQuery(
+  contract: Contract,
+  filter: EventFilter,
+  fromBlock: number,
+  toBlock: number,
+  originalBlock: number,
+  MAX_SEARCH_BLOCKS = 40000,
+  BLOCK_SPAN = 2048
+): Promise<{ recentLogs: Event[], blocksDelta: number }> {
+  if (originalBlock - fromBlock > MAX_SEARCH_BLOCKS) {
+    throw(new Error(`No events found within ${MAX_SEARCH_BLOCKS} blocks for ${contract.address}`));
+  }
+  try {
+    const res = await contract.queryFilter(filter, fromBlock, toBlock);
+    if (res.length > 0) {
+      return { recentLogs: res, blocksDelta: toBlock - fromBlock };
+    } else {
+      const nextToBlock = fromBlock;
+      const nextFrom = fromBlock - BLOCK_SPAN;
+      if (nextFrom < 0) {
+        throw(new Error('No events found by chain genesis'));
+      }
+      return fetchQuery(contract, filter, nextFrom, nextToBlock, originalBlock);
+    }
+  } catch (err) {
+    if (err.message.includes('query returned more')) {
+      const midBlock = (fromBlock + toBlock) / 2;
+      return fetchQuery(contract, filter, midBlock, toBlock, originalBlock);
```

Comment:
Missing `await`

Reply:
Those aren't missing... when you return `await` its the same as just returning the promise

Reply:
Oh interesting, TIL. Apparently there's still a slight difference. `return await` is slightly costlier but better for debugging since it keeps the function on the call stack so it shows up in the stack trace.

Reply:
word